### DISCUSSION TITLE
tootctl accounts fix-duplicates unable to fix-duplicates

### DISCUSSION THREAD

**Discussion by SMJ** [mastodon/mastodon#27877] [General]
### Steps to reproduce the problem

Run tootctl accounts fix-duplicates on a database with account duplicates with mastodon offline.
run reindex index accounts;
`ERROR:  could not create unique index "index_accounts_on_id"`

Use this query to find remaining duplicates:

```
select username, id, count(*) from accounts group by username, id having count(*) > 1;

  username   |         id         | count 
-------------+--------------------+-------
 a  | n |     2
 b  | n |     2
 c  | n |     2
 d  | n |     2
 e  | n |     2
(5 rows)

(duplicate usernames = alpha, id = n)
```

### Expected behaviour

fix-duplicates should be able to resolve duplicates.

### Actual behaviour

Duplicates remain and reindexing of the table is not possible.

### Specifications

Mastodon v4.0.2
ruby 2.7.2p137 
yarn 1.22.19
node v14.21.1

**Comment by SMJ** [comment]
Here is edited output (account url truncated) when running fix-duplicates:

`Error processing https://: PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_accounts_on_id"

Error processing https://: PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_accounts_on_id"

Error processing https://: PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_accounts_on_id"

Error processing https://: PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_accounts_on_id"

Error processing https://: PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_accounts_on_id"

Error processing https://: PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_accounts_on_id"

Error processing https://: PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_accounts_on_id"`

**Comment by SMJ** [comment]
When running accounts cull similar errors are seen.  This takes a couple of days to run to get to this point and seems to work through about 433,000 records before erroring out.

```
Error processing N: PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_accounts_on_id"
DETAIL:  Key (id)=(N) already exists.
Error processing N: PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_accounts_on_id"
DETAIL:  Key (id)=(N) already exists.                                                                                                                                      
Error processing N: PG::InternalError: ERROR:  tuple already updated by self                                                                                             
Error processing N: PG::InternalError: ERROR:  tuple already updated by self                                                                                             
Error processing N: PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_accounts_on_id"                                                               
DETAIL:  Key (id)=(N) already exists.

/home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/connection_adapters/postgresql_adapter.rb:687:in `exec_prepared': PG::DataCorrupted: ERROR:  missing chunk number 0 for toast value 6495995 in pg_toast_16504 (ActiveRecord::StatementInvalid)
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/connection_adapters/postgresql_adapter.rb:687:in `block (2 levels) in exec_cache'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.7/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.7/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.7/lib/active_support/dependencies/interlock.rb:47:in `permit_concurrent_loads'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/connection_adapters/postgresql_adapter.rb:686:in `block in exec_cache'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/connection_adapters/abstract_adapter.rb:696:in `block (2 levels) in log'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.7/lib/active_support/concurrency/load_interlock_aware_monitor.rb:26:in `block (2 levels) in synchronize'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.7/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.7/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.7/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.7/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/connection_adapters/abstract_adapter.rb:695:in `block in log'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.7/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/connection_adapters/abstract_adapter.rb:687:in `log'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/connection_adapters/postgresql_adapter.rb:685:in `exec_cache'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/connection_adapters/postgresql_adapter.rb:651:in `execute_and_clear'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/connection_adapters/postgresql/database_statements.rb:55:in `exec_query'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/connection_adapters/abstract/database_statements.rb:536:in `select_prepared'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/connection_adapters/abstract/database_statements.rb:67:in `select_all'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/connection_adapters/abstract/query_cache.rb:103:in `select_all'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/querying.rb:47:in `find_by_sql'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/relation.rb:843:in `block in exec_queries'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/relation.rb:858:in `block in skip_query_cache_if_necessary'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/connection_adapters/abstract/query_cache.rb:79:in `uncached'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/query_cache.rb:21:in `uncached'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/relation/delegation.rb:67:in `block in uncached'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/relation.rb:406:in `block in scoping'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/relation.rb:804:in `_scoping'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/relation.rb:406:in `scoping'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/relation/delegation.rb:67:in `uncached'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/relation.rb:857:in `skip_query_cache_if_necessary'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/relation.rb:828:in `exec_queries'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/relation.rb:631:in `load'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/relation.rb:249:in `records'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/relation/batches.rb:231:in `block in in_batches'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/relation/batches.rb:229:in `loop'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/relation/batches.rb:229:in `in_batches'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7/lib/active_record/relation/batches.rb:137:in `find_in_batches'
	from /home/mastodon/live/lib/mastodon/cli_helper.rb:40:in `parallelize_with_progress'
	from /home/mastodon/live/lib/mastodon/accounts_cli.rb:319:in `cull'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/thor-1.2.1/lib/thor/command.rb:27:in `run'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/thor-1.2.1/lib/thor/invocation.rb:127:in `invoke_command'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/thor-1.2.1/lib/thor.rb:392:in `dispatch'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/thor-1.2.1/lib/thor/invocation.rb:116:in `invoke'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/thor-1.2.1/lib/thor.rb:243:in `block in subcommand'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/thor-1.2.1/lib/thor/command.rb:27:in `run'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/thor-1.2.1/lib/thor/invocation.rb:127:in `invoke_command'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/thor-1.2.1/lib/thor.rb:392:in `dispatch'
	from /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/thor-1.2.1/lib/thor/base.rb:485:in `start'
	from bin/tootctl:8:in `<main>'
```

While a reindex table pg_toast.pg_toast_16504 does seem to complete, a reindex table accounts will continue to state that there are duplicate accounts.  At this point, it looks like careful surgery on the accounts table needs to be done.

**Comment by ClearlyClaire** [comment]
`fix-duplicates` is aimed at fixing duplicate records that are caused by text indexes behaving wrong… your case seems much more serious, with the primary key being duplicated?? this is not something you can recover from, you may need to restore a backup

**Comment by SMJ** [comment]
The duplicates look to be all remote accounts and on the test copy of the database I was able to remove statuses associated with account_id which then allowed me to remove the duplicate entry from accounts.  This particular account_id had been in the database since 2017.  

If I was to approach the other duplicates in the same way by removing either 1 or both from all tables and then rebuilding indexes, would there be any side affects?  It looks like account_id is used 47 times throughout the database for each account.

**Comment by ClearlyClaire** [comment]
Again, this is not possibly an issue with Mastodon itself, nor even an occurrence of the glibc issue that `fix-duplicates` is aimed at solving. Your database seem to include duplicate **primary identifiers**. This is not supposed to ever happen and it is a serious database corruption. This may be caused by a serious bug in PostgreSQL or another source of data corruption (e.g. faulty RAM).

> If I was to approach the other duplicates in the same way by removing either 1 or both from all tables and then rebuilding indexes, would there be any side affects? It looks like account_id is used 47 times throughout the database for each account.

If that is the only way your database is corrupted, then removing the duplicates should be enough, although I can't think of an automatic way to attribute related records to the “right” account, so I'd delete both. That being said, considering how seriously corrupted the database seems to be, I'm not sure it's wise to continue using it. Please have a look at https://wiki.postgresql.org/wiki/Corruption

**Comment by SMJ** [comment]
Thank you for the reply.  I'm working with a copy of the database and have made significant progress, it just seems the last step is to cleanly remove small list of remote account_ids that are duplicated.  If I could get some guidance for completely removing all rows associated with a single remote account_id I'd appreciate it.  Once this is done, I'll follow your recommendation of dumping all the data and then starting with a fresh database.

**Comment by ClearlyClaire** [comment]
Mastodon uses foreign keys with appropriate `ON DELETE CASCADE` and `ON DELETE SET NULL`, so deleting an offending account should not leave anything behind. However, I must again insist that as with all SQL queries, all bets are off with a corrupted database. Your database is not in a state that is inconsistent wrt. Mastodon itself, but inconsistent wrt. PostgreSQL guarantees, so operating like it were fine just runs the risk of making everything worse. Higher-level fixes should come after low-level fixes, not the other way around.

**Comment by SMJ** [comment]
Is bin/tootctl accounts fix-duplicates and rake mastodon:maintenance:find_duplicate_usernames the same thing?
If not, how can rake mastodon:maintenance:find_duplicate_usernames be called?

**Comment by christianp** [comment]
I've found this issue because our database is in the same situation, with some duplicate records in `accounts`. 
The duplicates all predate when PR #15364 was filed, so I assume we were affected by the bug fixed by that PR. Some of the duplicates have an empty `uri` field, but others have non-empty values.

**Comment by SMJ** [comment]
I'd be happy to share some notes with you of how I was able work this out.  Depending on how big your database is, you'll need some downtime chug through it all.  In some cases, the pg_toast corruption could be tracked down to a single column that contained non-critical data.  For duplicates choices can be made on what to keep.  Once you do the surgery, you can then rebuild indexes and repair toast.  Its best to setup a staging server to work through strategies, but it is possible to get through this and I'd be happy to help.
